src/relay/dnsrelay/mod.rs

```diff
@@ -1,194 +1,90 @@
-    #[cfg(target_os = "android")] local: &PathBuf,
-    #[cfg(not(target_os = "android"))] local: &SocketAddr,
-    remote: &Address,
-    qname: &Name,
-    qtype: RecordType,
-) -> io::Result<(Message, bool)> {
+    local: Arc<Local>,
+    remote: Arc<Remote>,
+    query: &Query
+) -> (io::Result<Message>, bool)
+    where
+        Local: upstream::Upstream,
+        Remote: upstream::Upstream,
+{
     // Start querying name servers
     debug!(
         "attempting lookup of {:?} {} with ns {:?} and {:?}",
-        qtype, qname, local, remote
+        query.query_type(), query.name(), local, remote
     );
 
-    // remove the last dot from fqdn name
-    let mut name = qname.to_ascii();
-    name.pop();
-    let addr = Address::DomainNameAddress(name, 0);
-    let qname_in_proxy_list = context.check_qname_in_proxy_list(&addr);
-
-    let remote_response_fut = async {
-        match qname_in_proxy_list {
-            Some(false) => None,
-            _ => {
-                let timeout = Some(Duration::new(3, 0));
-                try_timeout(proxy_lookup(context.clone(), svr_cfg, remote, qname, qtype), timeout)
-                    .await
-                    .ok()
-            }
-        }
-    };
-
-    let local_response = match qname_in_proxy_list {
-        Some(true) => None,
-        _ => {
-            let timeout = Some(Duration::new(3, 0));
-            try_timeout(local_lookup(qname, qtype, local), timeout).await.ok()
-        }
-    }
-    .unwrap_or_else(Message::new);
+    let remote_response_fut = try_timeout(remote.lookup(query), Some(Duration::new(3, 0)));
+    let local_response_fut = try_timeout(local.lookup(query), Some(Duration::new(3, 0)));
 
-    match qname_in_proxy_list {
+    match context.check_query_in_proxy_list(query) {
         Some(true) => {
-            let remote_response = remote_response_fut.await.unwrap_or_else(Message::new);
+            let remote_response = remote_response_fut.await;
             debug!("pick remote response (qname): {:?}", remote_response);
-            return Ok((remote_response, true));
+            return (remote_response, true);
         }
         Some(false) => {
+            let local_response = local_response_fut.await;
             debug!("pick local response (qname): {:?}", local_response);
-            return Ok((local_response, false));
+            return (local_response, false);
         }
         None => (),
     }
 
-    if local_response.answer_count() == 0 {
-        let remote_response = remote_response_fut.await.unwrap_or_else(Message::new);
-        return Ok((remote_response, true));
-    }
-
+    // FIXME: spawn(remote_response_fut)
+    let local_response = local_response_fut.await.unwrap_or_else(|_| Message::new());
     for rec in local_response.answers() {
+        if rec.record_type() != query.query_type() {
+            warn!("local DNS response has inconsistent answer type {} for query {}", rec.record_type(), query);
+            break
```

Comment:
This line looks problematic, it's possible we get `A` answers for  `CNAME` queries.

Reply:
A typical test case is `www.baidu.com`. Local DNS server returns `A` records even if query is `CNAME`, leading to wrong ACL routing, as we're assuming any domain name resolved by remote should be proxied.

Reply:
Okay, it's kind of DNS hijacking issue of my local ISP DNS server, but we'd better refine the ACL part to not proxy any hostname resolved by remote.

Reply:
Here is my fix: https://github.com/madeye/shadowsocks-rust/commit/98057a2e3ddfadffe57a762075759bec4ca1e8db

Reply:
```
$ dig @8.8.8.8 -t CNAME baidu.com

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> @8.8.8.8 -t CNAME baidu.com
; (1 server found)
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 37016
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;baidu.com.                     IN      CNAME

;; AUTHORITY SECTION:
baidu.com.              1799    IN      SOA     dns.baidu.com. sa.baidu.com. 2012142433 300 300 2592000 7200
```

To quote rfc1034:

> Answer          Carries RRs which _directly_ answer the query.
> ...
> Additional      Carries RRs which may be helpful in using the RRs in the
>                 other sections.

Under normal circumstances, only `A`/`AAAA`/`PTR` queries are issued.

P.S. An unpolluted DNS exchange with additional sections looks like --

```
$ dig -t MX google.com

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> -t MX google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 57293
;; flags: qr rd ra; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 7

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.com.                    IN      MX

;; ANSWER SECTION:
google.com.             600     IN      MX      50 alt4.aspmx.l.google.com.
google.com.             600     IN      MX      40 alt3.aspmx.l.google.com.
google.com.             600     IN      MX      30 alt2.aspmx.l.google.com.
google.com.             600     IN      MX      10 aspmx.l.google.com.
google.com.             600     IN      MX      20 alt1.aspmx.l.google.com.

;; ADDITIONAL SECTION:
alt2.aspmx.l.google.com. 152    IN      A       172.253.116.26
aspmx.l.google.com.     34      IN      A       172.217.197.26
alt1.aspmx.l.google.com. 177    IN      A       64.233.186.27
alt1.aspmx.l.google.com. 39     IN      AAAA    2800:3f0:4003:c00::1b
alt4.aspmx.l.google.com. 280    IN      A       172.217.218.26
alt3.aspmx.l.google.com. 280    IN      A       172.253.120.26
```

Reply:
tl;dr I believe this is fine.

Reply:
Okay rfc1034 does permit additional CNAME answers, which is a different issue.

```
   - The answer to the query, possibly preface by one or more CNAME
     RRs that specify aliases encountered on the way to an answer.
```

Reply:
Resolved via cb62eff.